### PR TITLE
[X86] Support branch hint

### PR DESCRIPTION
**Pull request by fzou1**
For more details about this feature, please refer to latest Intel 64 and IA-32 Architectures Optimization Reference Manual Volume 1: https://www.intel.com/content/www/us/en/content-details/821612/intel-64-and-ia-32-architectures-optimization-reference-manual-volume-1.html

### CODE REVIEW

File: llvm/lib/Target/X86/X86.td [hunk 1]
```diff
@@ -749,6 +749,11 @@ def TuningUseGLMDivSqrtCosts
     : SubtargetFeature<"use-glm-div-sqrt-costs", "UseGLMDivSqrtCosts", "true",
         "Use Goldmont specific floating point div/sqrt costs">;
 
+// Starting with Redwood Cove architecture, the branch has branch taken hint
+// (i.e., instruction prefix 3EH).
+def TuningBranchHint: SubtargetFeature<"branch-hint", "HasBranchHint", "true",
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86.td:754]: Why have you added this as a Tuning bit and not a Feature bit? Tuning bits are guaranteed to work on all applicable CPUs - they just might not be performant.

Comment (KanRobert) [reply llvm/lib/Target/X86/X86.td:754]: We can always add `0x3e` to JCC in 64-bit mode. So it does work on all applicable CPUs.

Comment (KanRobert) [reply llvm/lib/Target/X86/X86.td:754]: See Intel SDM
```
— Branch hints1:
• 2EH—Branch not taken (used only with Jcc instructions).
• 3EH—Branch taken (used only with Jcc instructions).

1. Some earlier microarchitectures used these as branch hints, but recent generations have not and they are reserved for future hint usage.
```

File: llvm/test/CodeGen/X86/branch-hint.ll [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+;       XX += 2;
+;   }
+;   for(int i=0; i<b; i++)
+;     bar(XX);
+; }
+
+@XX = dso_local local_unnamed_addr global i32 0, align 4
+
+; Function Attrs: nounwind
+define dso_local void @foo(i32 noundef %a, i32 noundef %b, i8* noundef readnone %c) local_unnamed_addr #0 {
```
Comment (KanRobert) [inline llvm/test/CodeGen/X86/branch-hint.ll:22]: Drop `dso_local` `local_unnamed_addr #0` `; Function Attrs: nounwind` and add `-mattr=+branch-hint` to run line.

File: llvm/test/CodeGen/X86/branch-hint.ll [hunk 3]
```diff
@@ -0,0 +1,95 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc < %s -mtriple=x86_64-linux -branch-hint -branch-hint-probability-threshold=60 -show-mc-encoding | FileCheck %s
+
+; Design: Insert "ds # encoding: [0x3e]" for condition branches who has high probability to take.
```
Comment (KanRobert) [inline llvm/test/CodeGen/X86/branch-hint.ll:4]: Maybe

`branches who has` -> `branch which has`?

Comment (KanRobert) [reply llvm/test/CodeGen/X86/branch-hint.ll:4]: `Insert "ds # encoding: [0x3e]"` -> `Add DS segment override prefix` ?

File: llvm/test/CodeGen/X86/branch-hint.ll [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+for.body:                                         ; preds = %if.end4, %for.body
+  %i.09 = phi i32 [ %inc, %for.body ], [ 0, %if.end4 ]
+  %1 = load i32, i32* @XX, align 4
+  %call = tail call i32 @bar(i32 noundef %1) #0
+  %inc = add nuw nsw i32 %i.09, 1
+  %exitcond.not = icmp eq i32 %inc, %b
+  br i1 %exitcond.not, label %for.cond.cleanup, label %for.body
+}
+
+declare dso_local i32 @bar(i32 noundef) local_unnamed_addr #0
```
Comment (KanRobert) [inline llvm/test/CodeGen/X86/branch-hint.ll:93]: drop `dso_local` `local_unnamed_addr #0`

File: llvm/test/CodeGen/X86/branch-hint.ll [hunk 5]
```diff
@@ -0,0 +1,95 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc < %s -mtriple=x86_64-linux -branch-hint -branch-hint-probability-threshold=60 -show-mc-encoding | FileCheck %s
```
Comment (KanRobert) [inline llvm/test/CodeGen/X86/branch-hint.ll:2]: Drop -show-mc-encoding

File: llvm/lib/Target/X86/X86MCInstLower.cpp [hunk 6]
```diff
@@ -54,6 +55,14 @@
 
 using namespace llvm;
 
+static cl::opt<bool> EnableBranchHint("branch-hint",
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86MCInstLower.cpp:58]: `enable-branch-hint` is more in line with our current naming habits

File: llvm/test/CodeGen/X86/branch-hint.ll
Comment (KanRobert) [inline llvm/test/CodeGen/X86/branch-hint.ll:1]: This test is too verbose, not robust and miss coverage.  I would suggest the following
```
; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
; RUN: llc < %s -mtriple=x86_64 -mattr=+branch-hint -branch-hint | FileCheck %s
; RUN: llc < %s -mtriple=x86_64 -mattr=+branch-hint -branch-hint -branch-hint-probability-threshold=50 | FileCheck %s
; RUN: llc < %s -mtriple=x86_64 -mattr=+branch-hint -branch-hint -branch-hint-probability-threshold=60 -tail-dup-placement=false | FileCheck --check-prefix=TH60 %s

define void @p51(i32 %x, ptr %p) {
; CHECK-LABEL: p51:
; CHECK:       # %bb.0: # %entry
; CHECK-NEXT:    testl %edi, %edi
; CHECK-NEXT:    ds
; CHECK-NEXT:    je .LBB0_2
; CHECK-NEXT:  # %bb.1: # %if.then
; CHECK-NEXT:    movl %edi, (%rsi)
; CHECK-NEXT:  .LBB0_2: # %if.end
; CHECK-NEXT:    retq
;
; TH60-LABEL: p51:
; TH60:       # %bb.0: # %entry
; TH60-NEXT:    testl %edi, %edi
; TH60-NEXT:    je .LBB0_2
; TH60-NEXT:  # %bb.1: # %if.then
; TH60-NEXT:    movl %edi, (%rsi)
; TH60-NEXT:  .LBB0_2: # %if.end
; TH60-NEXT:    retq
entry:
  %tobool.not = icmp eq i32 %x, 0
  br i1 %tobool.not, label %if.end, label %if.then, !prof !0

if.then:
  store i32 %x, ptr %p, align 4
  br label %if.end

if.end:
  ret void
}

define void @p61(i32 %x, ptr %p) {
; CHECK-LABEL: p61:
; CHECK:       # %bb.0: # %entry
; CHECK-NEXT:    testl %edi, %edi
; CHECK-NEXT:    jne .LBB1_1
; CHECK-NEXT:  # %bb.2: # %if.end
; CHECK-NEXT:    retq
; CHECK-NEXT:  .LBB1_1: # %if.then
; CHECK-NEXT:    movl %edi, (%rsi)
; CHECK-NEXT:    retq
;
; TH60-LABEL: p61:
; TH60:       # %bb.0: # %entry
; TH60-NEXT:    testl %edi, %edi
; TH60-NEXT:    ds
; TH60-NEXT:    je .LBB1_2
; TH60-NEXT:  # %bb.1: # %if.then
; TH60-NEXT:    movl %edi, (%rsi)
; TH60-NEXT:  .LBB1_2: # %if.end
; TH60-NEXT:    retq
entry:
  %tobool.not = icmp eq i32 %x, 0
  br i1 %tobool.not, label %if.end, label %if.then, !prof !1

if.then:
  store i32 %x, ptr %p, align 4
  br label %if.end

if.end:
  ret void
}
```

Comment (fzou1) [reply llvm/test/CodeGen/X86/branch-hint.ll:1]: Thanks. It's simpler. What's the metadata for "!prof !0" and "!prof !1"?

File: llvm/lib/Target/X86/X86.td (hunk 1 above)
Comment (fzou1) [reply llvm/lib/Target/X86/X86.td:754]: @KanRobert, Thank you for helping this.

File: llvm/test/CodeGen/X86/branch-hint.ll
Comment (KanRobert) [reply llvm/test/CodeGen/X86/branch-hint.ll:1]: Updated.

**Review by KanRobert — APPROVED** [review]
LGTM

File: llvm/lib/Target/X86/X86.td [hunk 7]
```diff
@@ -1815,12 +1822,12 @@ def : ProcModel<"pantherlake", AlderlakePModel,
 def : ProcModel<"clearwaterforest", AlderlakePModel,
                 ProcessorFeatures.CWFFeatures, ProcessorFeatures.ADLTuning>;
 def : ProcModel<"graniterapids", SapphireRapidsModel,
-                ProcessorFeatures.GNRFeatures, ProcessorFeatures.SPRTuning>;
+                ProcessorFeatures.GNRFeatures, ProcessorFeatures.GNRTuning>;
 def : ProcModel<"emeraldrapids", SapphireRapidsModel,
-                ProcessorFeatures.SPRFeatures, ProcessorFeatures.SPRTuning>;
+                ProcessorFeatures.SPRFeatures, ProcessorFeatures.GNRTuning>;
```
Comment (MattPD) [inline llvm/lib/Target/X86/X86.td:1827]: Isn't Emerald Rapids based on Raptor Lake microarchitecture (and not Redwood Cove)? If so, wouldn't branch hints be unsupported for this subtarget?

Comment (fzou1) [reply llvm/lib/Target/X86/X86.td:1827]: Nice catch. I'll correct it.

**Pull request merged**